proposals/esm-integration/README.md

```diff
@@ -88,57 +88,74 @@ Three things happen for each module during the Construction phase.
+
+WebAssembly exports can be any one of the [external types](https://webassembly.github.io/spec/core/exec/modules.html#external-typing), which currently include:
 
 - functions
 - tables
 - memories
 - globals
 
-Because WebAssembly does not currently have a type that can handle arbitrary JS values, JS value imports can not be handled with live bindings and would be snapshotted when they are imported. This means that in the short term, non-function imports coming from JS modules would be assigned a value of `undefined`, which will currently result in an error. 
+WebAssembly does not currently have a type that can handle arbitrary JS values. This means that JS value imports can not be handled with live bindings and will be snapshotted when they are imported. This means that if a JS module changes one of its exports to point to a different object, a WebAssembly module importing that object will not see the change.
 
-It may be possible to provide live bindings for JS values if an `any` type is added to WebAssembly. At this point the restriction prohibiting `undefined` imports could also be lifted. However, we would still want to throw an error for non-function imports in this case. 
+It may be possible to provide live bindings for JS values if an [`anyref`](https://github.com/WebAssembly/reference-types) type is added to WebAssembly.
 
-The reason we want to throw an error in this case is two fold:
+### Evaluation
 
-- **Reduce confusion for developers.** Since non-function imports work in JS->JS dependency chains, it is unintuitive that they would be undefined in JS->wasm dependency chains. Throwing an error makes clear what's happening.
-- **Allow for future support.** It may be possible to support non-function imports to wasm modules if an `any` type is added. We close off this avenue if we allow users to import values that can be undefined, though, as code may start depending on that behavior.
+During evaluation, the code is evaluated to assign values to the exported bindings. In JS, this means running the top-level module code.
 
-One important limitation of this approach to instantiation is that function imports won't work in JS in JS<->wasm cycles where the JS module is higher in the graph. In that case, the WebAssembly module is instantiated before the JS module, which means that functions haven't been initialized and would throw the same error as non-function imports. 
+WebAssembly has explicit types for both imports and exports. The type of exports is used to populate the lexical environment with a JavaScript value based on the WebAssembly export, from the [WebAssembly JS API](https://webassembly.github.io/spec/js-api/index.html). The type of imports is used to type check or convert imports to the required WebAssembly type.
```

Comment:
> WebAssembly has explicit types for both imports and exports.

It sounds a bit strange to me, could we remove that sentence? The following one implies that the export is typed, so I think it's understandable.

Reply:
> [...] based on the WebAssembly export

this seems duplicated.

Reply:
The reason I added this paragraph is because some of the below tables include behavior which is based on the import type, and others include behavior based on the export type. The previous version was all based on the export type, and it took me a while to decode the intentions. Many bytecode formats only include the types of exports and not imports, so this is a relevant detail.